### ISSUE TITLE
Predicted Probability in pmml of DecisionTreeClassifier

### ISSUE THREAD

**Issue by wayneleung555** [jpmml/sklearn2pmml#114]
Hi, vruusmann,

Below is the example in your github page, just wondering is there is way to generate the predicted probability in the output pmml ? I am trying to use the output "DecisionTreeIris.pmml" as a scoring code in SAS. Thanks.

https://github.com/jpmml/sklearn2pmml

```Python
import pandas
iris_df = pandas.read_csv("Iris.csv")
from sklearn.tree import DecisionTreeClassifier
from sklearn2pmml.pipeline import PMMLPipeline
from sklearn2pmml import sklearn2pmml
pipeline = PMMLPipeline([("classifier", DecisionTreeClassifier())])
pipeline.fit(iris_df[iris_df.columns.difference(["Species"])], iris_df["Species"])
sklearn2pmml(pipeline, "DecisionTreeIris.pmml", with_repr = True)
```

**Comment by vruusmann** [comment]
> is there is way to generate the predicted probability in the output pmml ?

What exactly do you mean by that? SkLearn2PMML generates a list of `<Model>/Output/OutputField@feature="probability"` elements for every classification-type model that implements the `ClassifierMixin.predict_proba(X)` method.

The `DecisionTreeClassifier` class is one of those models, so you shall find the following PMML markup in your "DecisionTreeIris.pmml" file (whether you asked for it or not):

```XML
<TreeModel>
	<Output>
		<OutputField name="probability(setosa)" optype="continuous" dataType="double" feature="probability" value="setosa"/>
		<OutputField name="probability(versicolor)" optype="continuous" dataType="double" feature="probability" value="versicolor"/>
		<OutputField name="probability(virginica)" optype="continuous" dataType="double" feature="probability" value="virginica"/>
	</Output>
</TreeModel>
```

**Comment by wayneleung555** [comment]
Hi, vruusmann,

Below is the pmml generated by running the above example, it does not contain probabilities ? I also added the pmml generated by package pmml in R at the end of this post, the probabilities (such as 0.907407407407407, 0.978260869565217) can been seen in the R pmml. Just wondering if i have missing anything ? Thanks. 





<?xml version="1.0" encoding="UTF-8" standalone="yes"?>
<PMML xmlns="http://www.dmg.org/PMML-4_3" xmlns:data="http://jpmml.org/jpmml-model/InlineTable" version="4.3">
	<Header>
		<Application name="JPMML-SkLearn" version="1.5.8"/>
		<Timestamp>2018-10-31T00:44:41Z</Timestamp>
	</Header>
	<MiningBuildTask>
		<Extension>PMMLPipeline(steps=[('classifier', DecisionTreeClassifier(class_weight=None, criterion='gini', max_depth=None,
            max_features=None, max_leaf_nodes=None,
            min_impurity_decrease=0.0, min_impurity_split=None,
            min_samples_leaf=1, min_samples_split=2,
            min_weight_fraction_leaf=0.0, presort=False, random_state=None,
            splitter='best'))])</Extension>
	</MiningBuildTask>
	<DataDictionary>
		<DataField name="Species" optype="categorical" dataType="string">
			<Value value="setosa"/>
			<Value value="versicolor"/>
			<Value value="virginica"/>
		</DataField>
		<DataField name="PetalLength" optype="continuous" dataType="float"/>
		<DataField name="PetalWidth" optype="continuous" dataType="float"/>
		<DataField name="SepalWidth" optype="continuous" dataType="float"/>
	</DataDictionary>
	<TransformationDictionary>
		<DerivedField name="double(PetalWidth)" optype="continuous" dataType="double">
			<FieldRef field="PetalWidth"/>
		</DerivedField>
		<DerivedField name="double(PetalLength)" optype="continuous" dataType="double">
			<FieldRef field="PetalLength"/>
		</DerivedField>
		<DerivedField name="double(SepalWidth)" optype="continuous" dataType="double">
			<FieldRef field="SepalWidth"/>
		</DerivedField>
	</TransformationDictionary>
	<TreeModel functionName="classification" missingValueStrategy="nullPrediction" splitCharacteristic="multiSplit">
		<MiningSchema>
			<MiningField name="Species" usageType="target"/>
			<MiningField name="PetalWidth"/>
			<MiningField name="PetalLength"/>
			<MiningField name="SepalWidth"/>
		</MiningSchema>
		<Output>
			<OutputField name="probability(setosa)" optype="continuous" dataType="double" feature="probability" value="setosa"/>
			<OutputField name="probability(versicolor)" optype="continuous" dataType="double" feature="probability" value="versicolor"/>
			<OutputField name="probability(virginica)" optype="continuous" dataType="double" feature="probability" value="virginica"/>
		</Output>
		<Node>
			<True/>
			<Node score="setosa" recordCount="50.0">
				<SimplePredicate field="double(PetalWidth)" operator="lessOrEqual" value="0.800000011920929"/>
				<ScoreDistribution value="setosa" recordCount="50.0"/>
				<ScoreDistribution value="versicolor" recordCount="0.0"/>
				<ScoreDistribution value="virginica" recordCount="0.0"/>
			</Node>
			<Node>
				<SimplePredicate field="double(PetalWidth)" operator="lessOrEqual" value="1.75"/>
				<Node>
					<SimplePredicate field="double(PetalLength)" operator="lessOrEqual" value="4.949999809265137"/>
					<Node score="versicolor" recordCount="47.0">
						<SimplePredicate field="double(PetalWidth)" operator="lessOrEqual" value="1.6500000953674316"/>
						<ScoreDistribution value="setosa" recordCount="0.0"/>
						<ScoreDistribution value="versicolor" recordCount="47.0"/>
						<ScoreDistribution value="virginica" recordCount="0.0"/>
					</Node>
					<Node score="virginica" recordCount="1.0">
						<True/>
						<ScoreDistribution value="setosa" recordCount="0.0"/>
						<ScoreDistribution value="versicolor" recordCount="0.0"/>
						<ScoreDistribution value="virginica" recordCount="1.0"/>
					</Node>
				</Node>
				<Node score="virginica" recordCount="3.0">
					<SimplePredicate field="double(PetalWidth)" operator="lessOrEqual" value="1.5499999523162842"/>
					<ScoreDistribution value="setosa" recordCount="0.0"/>
					<ScoreDistribution value="versicolor" recordCount="0.0"/>
					<ScoreDistribution value="virginica" recordCount="3.0"/>
				</Node>
				<Node score="versicolor" recordCount="2.0">
					<SimplePredicate field="double(PetalLength)" operator="lessOrEqual" value="5.449999809265137"/>
					<ScoreDistribution value="setosa" recordCount="0.0"/>
					<ScoreDistribution value="versicolor" recordCount="2.0"/>
					<ScoreDistribution value="virginica" recordCount="0.0"/>
				</Node>
				<Node score="virginica" recordCount="1.0">
					<True/>
					<ScoreDistribution value="setosa" recordCount="0.0"/>
					<ScoreDistribution value="versicolor" recordCount="0.0"/>
					<ScoreDistribution value="virginica" recordCount="1.0"/>
				</Node>
			</Node>
			<Node>
				<SimplePredicate field="double(PetalLength)" operator="lessOrEqual" value="4.850000381469727"/>
				<Node score="virginica" recordCount="2.0">
					<SimplePredicate field="double(SepalWidth)" operator="lessOrEqual" value="3.0999999046325684"/>
					<ScoreDistribution value="setosa" recordCount="0.0"/>
					<ScoreDistribution value="versicolor" recordCount="0.0"/>
					<ScoreDistribution value="virginica" recordCount="2.0"/>
				</Node>
				<Node score="versicolor" recordCount="1.0">
					<True/>
					<ScoreDistribution value="setosa" recordCount="0.0"/>
					<ScoreDistribution value="versicolor" recordCount="1.0"/>
					<ScoreDistribution value="virginica" recordCount="0.0"/>
				</Node>
			</Node>
			<Node score="virginica" recordCount="43.0">
				<True/>
				<ScoreDistribution value="setosa" recordCount="0.0"/>
				<ScoreDistribution value="versicolor" recordCount="0.0"/>
				<ScoreDistribution value="virginica" recordCount="43.0"/>
			</Node>
		</Node>
	</TreeModel>
</PMML>
 




###################################################
Extract from R package pmml as below : 
We can see probabilities of 0.907407407407407, 0.978260869565217 etc ........ : 



    <Node id="6" score="versicolor" recordCount="54">
     <CompoundPredicate booleanOperator="surrogate">
      <SimplePredicate field="Petal.Width" operator="lessThan" value="1.75"/>
      <SimplePredicate field="Petal.Length" operator="lessThan" value="4.75"/>
      <SimplePredicate field="Sepal.Length" operator="lessThan" value="6.15"/>
      <SimplePredicate field="Sepal.Width" operator="lessThan" value="2.95"/>
     </CompoundPredicate>
     <ScoreDistribution value="setosa" recordCount="0" confidence="0"/>
     <ScoreDistribution value="versicolor" recordCount="49" confidence="0.907407407407407"/>
     <ScoreDistribution value="virginica" recordCount="5" confidence="0.0925925925925926"/>
    </Node>
    <Node id="7" score="virginica" recordCount="46">
     <CompoundPredicate booleanOperator="surrogate">
      <SimplePredicate field="Petal.Width" operator="greaterOrEqual" value="1.75"/>
      <SimplePredicate field="Petal.Length" operator="greaterOrEqual" value="4.75"/>
      <SimplePredicate field="Sepal.Length" operator="greaterOrEqual" value="6.15"/>
      <SimplePredicate field="Sepal.Width" operator="greaterOrEqual" value="2.95"/>
     </CompoundPredicate>
     <ScoreDistribution value="setosa" recordCount="0" confidence="0"/>
     <ScoreDistribution value="versicolor" recordCount="1" confidence="0.0217391304347826"/>
     <ScoreDistribution value="virginica" recordCount="45" confidence="0.978260869565217"/>
    </Node>

**Comment by vruusmann** [comment]
So, the question is really that why SkLearn2PMML/JPMML-SkLearn do not initialize the `ScoreDistribution@probability` attribute?

The answer is that the `ScoreDistribution@probability` attribute is optional according to PMML schema.

The only required attribute is `ScoreDistribution@recordCount`. Once you have record counts, you can calculate the estimated probability distribution manually: `p(class) = recordCount of a class / sum of all recordCounts`.

Please note that the R's `pmml` package has initialized the `ScoreDistribution@confidence` attribute, not the `ScoreDistribution@probability`. Strictly speaking, they are two different things.

**Comment by wayneleung555** [comment]
Does it mean ScoreDistribution@probability is not available for all other models of SkLearn2PMML/JPMML-SkLearn at all ? How about Neural Network, Support Vector Machines , Generalised Liner model ? Do we need to calculate the probabilities manually from recordCounts as well ? 

In future,  is there plan to add ScoreDistribution@probability to SkLearn2PMML/JPMML-SkLearn ? I am in a project to investigate converting python model scoring code to be used in SAS. The probabilities is the main item to look for. It would be very helpful if the probability can be directly available.

**Comment by vruusmann** [comment]
The PMML standard specifies that the probability distribution of a classification is exposed to the end user via `<Model>/Output/OutputField@feature="probability"` elements as explained in my first comment.

The SkLearn2PMML/JPMML-SkLearn toolkit adds this to all model types, including logistic regressions, neural networks etc. Indeed, SVM is an important exception here, because SVM is a non-probabilistic classification method. You can distinguish between probabilistic and non-probabilistic Scikit-Learn classifiers by observing whether they declare a `predict_proba(X)` method or not.

The `ScoreDistribution@probability` attribute is some internal piece of information/documentation.

TL;DR: SkLearn2PMML/JPMML-SkLearn are generating 100% standards-compliant PMML documents. If you score those models using the JPMML-Evaluator library then you shall get a probability distribution with each prediction. However, I don not know if the SAS scoring engine is equally capable or not.

**Comment by vruusmann** [comment]
The PMML standard specifies that the probability distribution of a classification is exposed to the end user via `<Model>/Output/OutputField@feature="probability"` elements as explained in my first comment.

The SkLearn2PMML/JPMML-SkLearn toolkit adds this to all model types, including logistic regressions, neural networks etc. Indeed, SVM is an important exception here, because SVM is a non-probabilistic classification method. You can distinguish between probabilistic and non-probabilistic Scikit-Learn classifiers by observing whether they declare a `predict_proba(X)` method or not.

The `ScoreDistribution@probability` attribute is some internal piece of information/documentation.

TL;DR: SkLearn2PMML/JPMML-SkLearn are generating 100% standards-compliant PMML documents. If you score those models using the JPMML-Evaluator library then you shall get a probability distribution with each prediction. However, I don not know if the SAS scoring engine is equally capable or not.

**Comment by wayneleung555** [comment]
Is it possible to include ScoreDistribution@probability for both sklearn2pmml and pyspark2pmml  in future ?

**Comment by vruusmann** [comment]
> Is it possible to include ScoreDistribution@probability for both sklearn2pmml and pyspark2pmml in future ?

The `ScoreDistribution@recordCount` is a primary/required attribute; the `ScoreDistribution@probability` is a secondary/optional attribute.

Probabilities are calculated by performing a "simplemax" normalization over record counts. If the original PMML producer didin't do so, then you can always post-process the PMML file and calculate them yourself.

The SkLearn2PMML and PySpark2PMML packages are not pre-calculating them, because they serve no purpose (like explained above, the PMML consumer will calculate them on-the-fly if requested) and would only clutter the view.

**Comment by XScarlett** [comment]
> > Is it possible to include ScoreDistribution@probability for both sklearn2pmml and pyspark2pmml in future ?
> 
> The `ScoreDistribution@recordCount` is a primary/required attribute; the `ScoreDistribution@probability` is a secondary/optional attribute.
> 
> Probabilities are calculated by performing a "simplemax" normalization over record counts. If the original PMML producer didin't do so, then you can always post-process the PMML file and calculate them yourself.
> 
> The SkLearn2PMML and PySpark2PMML packages are not pre-calculating them, because they serve no purpose (like explained above, the PMML consumer will calculate them on-the-fly if requested) and would only clutter the view.

Hi vruusmann,
Due to business requirements, The model training uses sample_weights , so the recordCount is not correct, in this case, how to calculate **probability** in pmml?
The tree recordCount like this:
![image](https://user-images.githubusercontent.com/30850280/83618414-ff119b00-a5bc-11ea-9fba-2beb80e79afa.png)
